lib/multierror/multierror.go

```diff
@@ -46,13 +59,16 @@ func NewOr(errs []error, fallback error) error {
 // Unwrap for MultiError always returns nil, as there is no reasonable way to implement it.
 //
 // Use As and Is methods, or loop over the list directly to access the underlying errors.
-func (me MultiError) Unwrap() error {
-	return nil
+func (multi MultiError) Unwrap() error {
+	if len(multi) == 0 {
```

Comment:
Why always the first error?

Reply:
I went back because of this comment, and realized I had implemented it wrong! :> now it properly unwraps

Reply:
The reason this was not implemented is that semantically the errors in a MultiError are NOT nested.

Let me expand on this. Let's say you get a DataBaseError. You Unwrap it, and discover that you got a DatabaseError because there was a ConnectionError. You Unwrap it again, and discover that you got a ConnectionError because of a DNSResolutionError.
In the code, you can say "if DataBaseError IS a DNSResolutionError", then ... retry without DNS, for example.

Now, let's say that I now have a MultiError containing 3 DataBaseError. The caller invokes Unwrap(). The current code returns just another MultiError, with 2 DatabaseError. Instead of returning THE CAUSE of an error, it just moves to the next element in the array. It then calls Unwrap() again, and another MultiError with 1 element is returned. So again, not the cause of the error. Finally, nil is returned.

Can you add a test with error nesting, and see what happens with Is and As? It's hard to follow the code in the golang library, but my guess is that the change in this PR are semantically breaking the expected behavior of Unwrap.

Can you try a test with normally nested errors? I suspect this test would pass BEFORE your change, but would FAIL after your change:
```
  errs := []error{
    fmt.Errorf("outer caused by %w", fmt.Errorf("inner caused by %w", OneErr)),
    fmt.Errorf("outer caused by %w", fmt.Errorf("inner caused by %w", TwoErr)),
  }

  assert.True(t, errors.Is(errs, OneErr))
  assert.True(t, errors.Is(errs, TwoErr))
  assert.False(t, errors.Is(errs, ThreeErr))
```

Reply:
(please don't merge unless this comment is addressed - until I'm convinced otherwise, while I think the tests are valuable, I think this is breaking the contract for Unwrap(), Is() and As())

Reply:
1. I already have nested testing [here](https://github.com/enfabrica/enkit/pull/355/files#diff-b0d1f1c620fbee8ca09faef28317d5a966fb36cc07eefd0808b8f51a5ab65cf5R68) I also added in your example.
2. The code snippet you passed in passes both before and after. These changes as a whole are intentionally not behavior changing. That's why I made tests.
3. What the current library does is call errors.Is on all of it's existing errors, instead of letting the errors.Is function body itself do that.

Reply:
The reason the code above works is that it relies on `errors.Is` to call `Unwrap()` when the current error doesn't match, which effectively recursively replaces the loop in the old `multierror.Is` implementation.

The debate is on whether `multierror.Unwrap` iterating over errors makes sense. `Unwrap()` is supposed to reveal the cause of errors; in the case of multierror, the next error is probably not the cause of the previous error, so this will be confusing to multierror users.

There are useful implementations of `As()` and `Is()` for `multierror`, as previously shown. Its clear that this code can be refactored by implementing `Unwrap()` (which `errors.As` and `errors.Is` leverages), but `Unwrap()` by itself does something different than it should (iterating over disparate errors rather than finding the cause of any particular error) and I think that's the main objection.

Reply:
This would be true if, and only if, Unwrap actually had a reason to exist. It does not. Referenced multiple times in the errors package, like [here](https://cs.opensource.google/go/go/+/refs/tags/go1.17.5:src/errors/wrap.go;l=26), Unwrap is simply a pop from a list. errors.As and errors.Is serve that functionality to "reveal" causes of errors.

To have a list of errors, and when called to Unwrap (aka to pop) and return nil even though the end user knows there are multiple errors in there is contract breaking.

Reply:
What if, for example, I wrote a errors library that will parse the typing of each err in an err chain and upload it to some statistics database. 

I would not use errors.As or errors.Is, I would use Unwrap to parse all individual errors and their typing. So I would expect that if err != nil for a top level operation, Unwrap shouldn't return nothing.